### PR TITLE
Enable `tie_word_embeddings` config setting to enable / disable weight tied embeddings

### PR DESCRIPTION
**Pull request by vchiley**
Enable disabling embedding weight tying
(Using huggingface `PretrainedConfig`'s `tie_word_embeddings` kwarg)

Same seed, same config (except `tie_word_embeddings: {true, false}`), produces slightly different convergence curves
Screenshot 2023-11-09 at 5 04 05 PM
and the resulting param count is different:
Screenshot 2023-11-09 at 5 06 02 PM

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -574,14 +575,20 @@ class MPTForCausalLM(MPTPreTrainedModel):
...
-        if not config.tie_word_embeddings:
-            raise ValueError(
-                'MPTForCausalLM only supports tied word embeddings')
-
         log.info(f'Instantiating an MPTForCausalLM model from {__file__}')
 
         self.transformer: MPTModel = MPTModel(config)
 
+        self.lm_head = None
+        if config.tie_word_embeddings is False:
```
Comment (sashaDoubov) [inline llmfoundry/models/mpt/modeling_mpt.py:583]: was there a reason to do `if config.tie_word_embeddings is False:` instead of `if not config.tie_word_embeddings`:

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -602,19 +609,30 @@ def __init__(self, config: MPTConfig):
...
     def set_input_embeddings(
             self, value: Union[SharedEmbedding, nn.Embedding]) -> None:
-        self.transformer.wte = value
+        self.transformer.set_input_embeddings(value)
 
-    def get_output_embeddings(self) -> nn.Embedding:
-        return self.transformer.wte
+    def get_output_embeddings(
+            self) -> Union[SharedEmbedding, nn.Embedding, nn.Linear]:
+        return self.lm_head or self.transformer.get_input_embeddings()
```
Comment (sashaDoubov) [inline llmfoundry/models/mpt/modeling_mpt.py:621]:
```suggestion
        if self.lm_head is not None:
            return self.lm_head
        return self.transformer.get_input_embeddings()
```
would prefer this to be more explicit

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -859,7 +879,11 @@ def flops_per_batch(self, batch: Mapping) -> int:
         # assume the backward pass is approximately 2x the forward pass
 
         bs, msl = batch['input_ids'].shape[0:2]
-        params_flops_per_token = 2 * self.n_active_params
+        params = self.n_active_params
+        if self.model.transformer.config.tie_word_embeddings is False:
```
Comment (sashaDoubov) [inline llmfoundry/models/mpt/modeling_mpt.py:883]: same nit.

Comment (sashaDoubov) [inline llmfoundry/models/mpt/modeling_mpt.py:883]: (potentially) same nit, this is up to you.

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 1 above)
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:583]: I do this because it seems more explicit, but 🤷‍♂️
I'll change it

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 4]
```diff
@@ -602,19 +609,32 @@ def __init__(self, config: MPTConfig):
...
 
     def set_output_embeddings(
-            self, new_embeddings: Union[SharedEmbedding, nn.Embedding]) -> None:
-        self.transformer.wte = new_embeddings
+        self, new_embeddings: Union[SharedEmbedding, nn.Embedding,
+                                    nn.Linear]) -> None:
+        if self.lm_head is not None:
+            self.lm_head = new_embeddings
+        else:
+            assert isinstance(new_embeddings, (SharedEmbedding, nn.Embedding))
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:631]: this should probably be an actual error since it is invalid input, not just convincing pyright?

File: llmfoundry/models/mpt/modeling_mpt.py [hunk 5]
```diff
@@ -602,19 +609,32 @@ def __init__(self, config: MPTConfig):
...
+            self.transformer.set_input_embeddings(new_embeddings)
+
+    def tie_weights(self) -> None:
+        if self.lm_head is not None:
+            del self.lm_head
```
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:636]: Is the `del` necessary?

File: tests/test_model.py [hunk 6]
```diff
@@ -796,10 +805,15 @@ def test_generate(attention_impl: str, device: str, pos_emb_config: dict):
...
             **pos_emb_config,
         },
+        tie_word_embeddings=tie_word_embeddings,
     )
     mpt = MPTForCausalLM(hf_config)
-    mpt.eval()
+    if not tie_word_embeddings:
+        assert mpt.lm_head is not None
+        with torch.no_grad():
+            mpt.lm_head.weight.copy_(mpt.transformer.wte.weight)
```
Comment (dakinggg) [inline tests/test_model.py:814]: why do you need this?

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:636]: probably not, was just being a little too thorough.
will remove

File: tests/test_model.py (hunk 6 above)
Comment (vchiley) [reply tests/test_model.py:814]: I think test didn't pass without it.
My reasoning: given the tokens are hardcoded, the init of the lm_head needed to be the same as the input embed to match the hardcoded tokens.

File: llmfoundry/models/mpt/modeling_mpt.py (hunk 4 above)
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:631]: Initially, I did include it just to get past pyright, but can make it more of a real error.

**Review by dakinggg — APPROVED** [review]

File: tests/test_model.py (hunk 6 above)
Comment (dakinggg) [reply tests/test_model.py:814]: wdym "the tokens are harcoded"? I think only the input is?

Comment (vchiley) [reply tests/test_model.py]: found this was caused by a precision / sampling issue.
adding `precision` arg to test and skipping test that fails.

File: tests/test_model.py [hunk 7]
```diff
@@ -1237,7 +1259,7 @@ def test_forward_with_cache(attn_impl: str, device: str, pos_emb_config: dict):
         torch.testing.assert_close(
             second_output.logits,
             full_output.logits[:, -1, :].unsqueeze(1),
-            atol=1e-2,
+            atol=1e-1,
```
Comment (dakinggg) [inline tests/test_model.py:1262]: what about 1.1e-2

Comment (vchiley) [reply tests/test_model.py:1262]: that seems to pass as well.

**Review by dakinggg — APPROVED** [review]
